File: services/internal/domain/event_dedupe.go
package domain

type FlightEventDedupeKeyInput struct {
	EventType              FlightEventType
	FAFlightID             *FAFlightID
	ProvisionalFlightLegID *ProvisionalFlightLegID
	EventTimestamp         ISODateTimeString
}

func GenerateFlightEventDedupeKey(input FlightEventDedupeKeyInput) string {
	return hashFlightIDParts(
		string(FlightEventSourcePolling),
		string(input.EventType),
		eventDedupeFlightKey(input),
		string(input.EventTimestamp),
	)
}

func eventDedupeFlightKey(input FlightEventDedupeKeyInput) string {
	if input.FAFlightID != nil {
		return string(*input.FAFlightID)
	}
	if input.ProvisionalFlightLegID != nil {
		return string(*input.ProvisionalFlightLegID)
	}

	panic("faFlightId or provisionalFlightLegId is required for event dedupe keys")
}

File: packages/shared-types/src/event-dedupe.ts
import type {
  FAFlightID,
  FlightEventType,
  ISODateTimeString,
  ProvisionalFlightLegID,
} from "./domain-types.js";
import { hashStableParts } from "./stable-hash.js";

export interface FlightEventDedupeKeyInput {
  eventType: FlightEventType;
  faFlightId: FAFlightID | null;
  provisionalFlightLegId: ProvisionalFlightLegID | null;
  eventTimestamp: ISODateTimeString;
}

export function generateFlightEventDedupeKey(input: FlightEventDedupeKeyInput): string {
  return hashStableParts([
    "polling",
    input.eventType,
    requireFlightKey(input.faFlightId, input.provisionalFlightLegId),
    input.eventTimestamp,
  ]);
}

function requireFlightKey(
  faFlightId: FAFlightID | null,
  provisionalFlightLegId: ProvisionalFlightLegID | null,
): string {
  const flightKey = faFlightId ?? provisionalFlightLegId;
  if (flightKey === null) {
    throw new Error("faFlightId or provisionalFlightLegId is required for event dedupe keys");
  }

  return flightKey;
}
